zkevm-circuits/src/evm_circuit/execution.rs

```diff
@@ -798,26 +839,19 @@ impl<F: Field> ExecutionConfig<F> {
                     }
 
                     offset += height;
-                    last_height = height;
 
-                    if step.execution_state == ExecutionState::EndBlock {
-                        // evm_circuit_pad_to == 0 means no extra padding
-                        if exact || block.evm_circuit_pad_to == 0 {
-                            // no padding
-                            break;
-                        } else {
-                            // padding
-                            if offset >= block.evm_circuit_pad_to {
-                                if offset > block.evm_circuit_pad_to {
-                                    log::warn!(
-                                        "evm circuit offset larger than padding: {} > {}",
-                                        offset,
-                                        block.evm_circuit_pad_to
-                                    );
-                                }
-                                break;
-                            }
-                        }
+                    if !exact && offset >= evm_rows {
```

Comment:
@ed255 Off by one or intended?

Reply:
ah this is tricky, but I think this is not a bug.
If there are 10 rows, the offset can only go from 0 to 9, so offset=10 should trigger the error.
The log message is wrong though, it should be `evm circuit offset larger than padding: {} >= {}`

Reply:
The offset calculation is a bit brittle and I attempt to fix it via #870 . I encountered a few more issues depending on circuit configuration and the number of transactions in geth data etc

Reply:
Actually I checked it again and it was an off by one bug T.T  This is because at this line, the offset contains the next available offset.  It's OK for the next available offset to be equal to evm_rows, as long as we finish the loop here.  So it should be `offset > evm_rows` in the check.

I saw your PR!  Thanks for the effort on simplifying the offset handling logic!  I still think there are some things that need fixing that I completely missed on my EndBlock PR.